cmd/kubeapps-apis/plugins/helm/packages/v1alpha1/repositories.go

```diff
@@ -0,0 +1,146 @@
+	corev1 "github.com/vmware-tanzu/kubeapps/cmd/kubeapps-apis/gen/core/packages/v1alpha1"
+	"github.com/vmware-tanzu/kubeapps/cmd/kubeapps-apis/plugins/pkg/statuserror"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	k8scorev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/utils/strings/slices"
+)
+
+const (
+	HelmRepoType   = "helm"
+	OCIRepoType    = "oci"
+	RedactedString = "REDACTED"
+)
+
+func (s *Server) newRepo(ctx context.Context, targetName types.NamespacedName, url string, repoType string, description string, interval uint32,
+	tlsConfig *corev1.PackageRepositoryTlsConfig, auth *corev1.PackageRepositoryAuth) (*corev1.PackageRepositoryReference, error) {
+	if url == "" {
+		return nil, status.Errorf(codes.InvalidArgument, "repository url may not be empty")
+	}
+	if repoType == "" || !slices.Contains(getValidRepoTypes(), repoType) {
+		return nil, status.Errorf(codes.InvalidArgument, "repository type [%s] not supported", repoType)
+	}
+
+	var secret *k8scorev1.Secret
+	typedClient, err := s.clientGetter.Typed(ctx, s.kubeappsCluster)
+	if err != nil {
+		return nil, err
+	}
+	if s.pluginConfig.UserManagedSecrets {
+		if secret, err = validateUserManagedRepoSecret(ctx, typedClient, targetName, tlsConfig, auth); err != nil {
+			return nil, err
+		}
+	} else {
+		// a bit of catch 22: I need to create a secret first, so that I can create a repo that references it
+		// but then I need to set the owner reference on this secret to the repo. In has to be done
+		// in that order because to set an owner ref you need object (i.e. repo) UID, which you only get
+		// once the object's been created
+		if secret, err = createKubeappsManagedRepoSecret(ctx, typedClient, targetName, tlsConfig, auth); err != nil {
+			return nil, err
+		}
+	}
+
+	passCredentials := auth != nil && auth.PassCredentials
+	insecureSkipVerify := tlsConfig != nil && tlsConfig.InsecureSkipVerify
+
+	if helmRepoCrd, err := newHelmRepoCrd(targetName, url, repoType, description, interval, secret, tlsConfig, auth, passCredentials, insecureSkipVerify); err != nil {
+		return nil, err
+	} else if client, err := s.getClient(ctx, targetName.Namespace); err != nil {
+		return nil, err
+	} else if err = client.Create(ctx, helmRepoCrd); err != nil {
+		return nil, statuserror.FromK8sError("create", "AppRepository", targetName.String(), err)
+	} else {
+		if !s.pluginConfig.UserManagedSecrets {
+			if err = s.setOwnerReferencesForRepoSecret(ctx, secret, helmRepoCrd); err != nil {
+				return nil, err
+			}
+		}
+		return &corev1.PackageRepositoryReference{
+			Context: &corev1.Context{
+				Namespace: helmRepoCrd.Namespace,
+				Cluster:   s.kubeappsCluster,
+			},
+			Identifier: helmRepoCrd.Name,
+			Plugin:     GetPluginDetail(),
+		}, nil
+	}
+}
+
+func getValidRepoTypes() []string {
```

Comment:
could that be a constant?

Reply:
It seems that arrays in Go can't be constants. [See here.](https://stackoverflow.com/a/13137568)

Reply:
good to know 👍